core/primitives/src/receipt.rs

```diff
@@ -127,6 +127,26 @@ impl Receipt {
 pub enum ReceiptEnum {
     Action(ActionReceipt),
     Data(DataReceipt),
+    PromiseYield(ActionReceipt),
+    PromiseResume(DataReceipt),
```

Comment:
This enum is borsh-serializable. Does the serialized payload end up on the chain? If so, are there any considerations we need to keep in mind for compatibility here?

In particular I would probably specify the discriminants explicitly at least. This also now should have explicit `Borsh(De)Serialize` implementations that make sure to not handle these variants in any way before the feature is stable to not accidentally change the observed behaviour (in particular, note that this includes handling of any previously invalid data that might have become valid after this change.)

Reply:
Thanks, yes, this area is worth some attention but I wonder if it actually necessary to do anything further. Per [borsh specification](https://github.com/near/borsh?tab=readme-ov-file#specification), enums are serialized starting with the enum ordinal, so adding new variants to an existing enum shouldn't break FC/BC. I also see examples such as [Block enum](https://github.com/near/nearcore/blob/7427cfcb5503f3db0b7ce474e1fd524dfe98083b/core/primitives/src/block.rs#L83-L91) which is borsh-serializable, we add new variants over time, and there is no need for explicit serialize/deserialize implementations.

Reply:
I'm not too worried about the schema compatibility, but rather the fact that we should not currently accept or interpret in any way any receipts from the network with tags 2 (PromiseYield) or 3 (PromiseResume). In normal operation where all nodes are well behaved and use the reference implementation this would not be a concern, but there is nothing stopping anybody from modifying the reference code such that there is some traffic of these currently malformed receipts. Until yield/resume is stable other well behaved nodes should not be able to pretend that those malformed messages are valid.

Reply:
(The alternative would be to `#[cfg(feature = "nightly_feature_promise_yield_resume")] or some such, that would also work for me)

Reply:
I've gated the deserialization, could you please take a look? Do you think the serialization needs to be gated as well?